### PR TITLE
[#200] Django Wiki

### PR DESCRIPTION
**Pull request by gdude2002**
This is an early PR just to keep track of the progress on the wiki. I'll update the title when it's ready to go!

Some notes:

* The wiki will be at `/wiki` on the base domain, not a `wiki` subdomain. This overall seems to be preferential anyway, but I was also having trouble actually getting `django-wiki` to work at all on another host.
* I've included the Monokai Pygments theme for code block highlighting. The main reason for this is that it might be useful to have it accessible to other parts of the site - not just the wiki.
* This could be a long PR - there's a lot of template work to do. It may or may not also be blocked by the lack of a working login system, but we'll see how we go.

### CODE REVIEW

**Comment by gdude2002** [comment]
And with that, the PR is ready for review!

**Comment by gdude2002** [comment]
There's quite a lot of stuff in this PR, so here's a summary:

* Changes to `settings.py`:
    * Added a bunch of stuff that the wiki requires
    * `LOGIN_URL` - this will need to be edited when an auth system is in place
    * Updated `PARENT_HOST` env var so it can be used in DEBUG mode and will be added to `ALLOWED_HOSTS` in that scenario
    * A few accessibility changes to the Bulma variables
    * Django's media storage has been set up - that'll be found under `media/` in the root
* Wiki
    * Fully implemented
    * There are a few things that don't look fully polished - these are purely on the backend, and can be addressed later if needs be
    * Let me tell ya: This thing was a damn pain. Lots of weird decisions. Lots of odd code. But it works!
        * [me_irl](https://youtu.be/rvrZJ5C_Nwg?t=127)
* Dependencies
    * Re-locked the Pipfile, since it had to be merged
    * Added `wiki` and `pygments`

File: pydis_site/settings.py
Comment (lemonsaurus) [inline pydis_site/settings.py]: Not sure it's necessary to justify this change with an inline comment.

File: pydis_site/settings.py
Comment (lemonsaurus) [inline pydis_site/settings.py:283]: not a big fan of these mid-dict newlines. I mean, green is a color, the previous two are colors, they shouldn't even be in separate groups as far as I can reason?

File: pydis_site/settings.py
Comment (lemonsaurus) [inline pydis_site/settings.py:313]: this newline, however, seems fine. logical groupings of 5+ items like that. it just looks weird in the other one, though.

File: pydis_site/settings.py
Comment (lemonsaurus) [inline pydis_site/settings.py:323]: wait, so what does this do exactly? How will debug messages look when there's nothing here?

File: pydis_site/static/js/wiki/dropdown.js
Comment (lemonsaurus) [inline pydis_site/static/js/wiki/dropdown.js]: Maybe instead of leaving this comment in, you could make an issue over on that repo and just stick this code right in the issue?

File: LICENSE-SimpleMDE
Comment (jchristgit) [inline LICENSE-SimpleMDE:21]: this is from the markdown editor, right? shoudln't it reside in the static/js folder instead of top-level? otherwise we will probably end up with dozens of licenses t some point

File: pydis_site/templates/base/base.html
Comment (lemonsaurus) [inline pydis_site/templates/base/base.html]: I think I prefer the readability of splitting this into three loads. I feel like the python equivalent to this would be 

```py
import django_simple_bulma, sekizai_tags, static
```

and we discourage that, for good reasons.

File: pydis_site/templates/wiki/base.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/base.html]: same applies here - multiple loads will be more readable.

File: Pipfile
Comment (jchristgit) [inline Pipfile]: Not a fan of this at all. I'm also not sure if this can break things for us if we port back up to 2.2 later. Also, 2.2 is LTS, so it's a good version to stay on.

Looks like django-wiki depends on < 2.2, but I'm not sure if this is an actual "it breaks on 2.2" or "we didn't update the version specifier yet" - judging from [this commit](https://github.com/django-wiki/django-wiki/commit/018074b62800e2ec91abc6e3cb36f73bc1d4f6c1) they didn't add it to the test matrix yet and maybe could theoretically support it. Maybe we can send in a PR to add 2.2 to the test matrix? I can look into that later today.

File: pydis_site/apps/home/templatetags/wiki_extra.py
Comment (jchristgit) [inline pydis_site/apps/home/templatetags/wiki_extra.py]: do you need the extra `isinstance` check here? I feel like this entire block could just be
```py
unbound_field = get_unbound_field(field)
```
instead.

File: pydis_site/templates/wiki/base.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/base.html]: FYI: you may need to remove these on a lot of templates because in order to do a push-footer thing where it would always be at the bottom of the screen even if the content doesn't fill the page, I moved the footer to the base template in my branch.

File: pydis_site/settings.py
Comment (jchristgit) [inline pydis_site/settings.py:306]: this should probably be on a newline

File: pydis_site/templates/wiki/base.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/base.html:44]: any particular reason these are at the end of the file and not at the start? I know I've seen people do that as a sort of hack to overcome load-order related problems but if at all possible I'd love to see them in the head where they belong.

File: pydis_site/static/js/wiki/dropdown.js
Comment (jchristgit) [inline pydis_site/static/js/wiki/dropdown.js]: this should probably be an issue on the django-simple-bulma repo?

File: pydis_site/templates/wiki/base_site.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/base_site.html]: okay.. the rest seem to be done. maybe it's time to remove it?

File: pydis_site/templates/wiki/base_site.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/base_site.html]: nice. I can tell this one wasn't meant to be in the PR. :stuck_out_tongue:

File: pydis_site/templates/wiki/delete.html
Comment (jchristgit) [inline pydis_site/templates/wiki/delete.html:26]: shouldn't these checks go over django's permission system or how does this work?

File: pydis_site/templates/wiki/includes/breadcrumbs.html
Comment (jchristgit) [inline pydis_site/templates/wiki/includes/breadcrumbs.html:52]: i imagine not everybody should be able to create articles later? so we should make this an issue

File: pydis_site/templates/wiki/includes/render.html
Comment (jchristgit) [inline pydis_site/templates/wiki/includes/render.html]: same thing with multiline loads here

**Review by jchristgit — DISMISSED** [review]
- i think the load template tags should be on separate lines for git blame retention and other reasons but would be curious what others think

File: pydis_site/templates/wiki/create.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/create.html]: multi-line loads please.

File: pydis_site/templates/wiki/create_root.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/create_root.html]: multi-line loads please.

File: pydis_site/templates/wiki/delete.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/delete.html]: I'll stop mentioning these now, you'll obviously do all of them.

File: pydis_site/templates/wiki/history.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/history.html]: I'd prefer if this PR didn't have `<style>` tags either, much like how the js was moved into files, I suggest you do the same with the css.

File: pydis_site/templates/wiki/plugins/links/sidebar.html
Comment (lemonsaurus) [inline pydis_site/templates/wiki/plugins/links/sidebar.html]: this comment is weird. Let's just agree that this is fine.

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
Whew. Big PR, but for the most part looks like high quality stuff.

If you've got that demo still running, feel free to put that link in a comment so people can play with it a little as they review. A couple of screenshots may have been nice for a PR this big, too.

File: LICENSE-SimpleMDE
Comment (gdude2002) [reply LICENSE-SimpleMDE:21]: I think we probably need a place for them, by convention they go in the root but if you can propose somewhere we could collect them, that'd work

File: Pipfile
Comment (gdude2002) [reply Pipfile]: I think there isn't much chance of getting that merged, but if you're hopeful, be my guest.

Could delay the PR quite a bit if we're waiting on it, though.

File: pydis_site/templates/wiki/delete.html
Comment (gdude2002) [reply pydis_site/templates/wiki/delete.html:26]: They do. That's a wiki thing.

File: pydis_site/settings.py
Comment (gdude2002) [reply pydis_site/settings.py]: It's not an extremely different shade - it would feel weird to _not_ justify it, imo

File: pydis_site/settings.py
Comment (gdude2002) [reply pydis_site/settings.py:323]: They use the default grey colour.

image

image

File: pydis_site/templates/wiki/base.html
Comment (gdude2002) [reply pydis_site/templates/wiki/base.html:44]: That's the first reason people do it - the other is so that scripts that don't need to be loaded for the page to function don't block the rendering of the page.

We can use a JS event, though. It just means the page will take longer to render.

File: pydis_site/templates/wiki/base_site.html
Comment (gdude2002) [reply pydis_site/templates/wiki/base_site.html]: Haha, wasn't even my comment. I'll get it.

File: pydis_site/templates/wiki/base.html
Comment (gdude2002) [reply pydis_site/templates/wiki/base.html]: As long as your branch still has the block - I need to be able to remove the footer in the preview iframes

File: pydis_site/templates/wiki/includes/breadcrumbs.html
Comment (gdude2002) [reply pydis_site/templates/wiki/includes/breadcrumbs.html:52]: #201

File: pydis_site/templates/wiki/history.html
Comment (gdude2002) [reply pydis_site/templates/wiki/history.html]: Actually, this isn't used, so I'll just remove it.

File: LICENSE-SimpleMDE
Comment (gdude2002) [reply LICENSE-SimpleMDE:21]: As discussed on Discord, we're going to solve this one when it becomes a problem.

File: pydis_site/templates/wiki/base.html
Comment (lemonsaurus) [reply pydis_site/templates/wiki/base.html]: I'm not sure it does, but it's been merged now so you can find out.

File: pydis_site/templates/wiki/base.html
Comment (lemonsaurus) [reply pydis_site/templates/wiki/base.html:44]: naw, fuck it, leave it as is.

File: Pipfile
Comment (gdude2002) [reply Pipfile]: From Discord: We're not gonna wait to solve this before merging the PR, but it's something that should be prioritised after.

Comment (lemonsaurus) [reply Pipfile]: huh? which discord is this from? If you were paraphrasing me, you may have misunderstood. I absolutely do want Django 2.2 before we merge this, but I won't wait for months for the django-wiki people to merge a PR, though. If we have to, we will temporarily fork the wiki and change it ourselves.

Comment (gdude2002) [reply Pipfile]: Ah okay, fair

**Comment by gdude2002** [comment]
Reviews addressed and latest changes from `django` have been merged.

image

**Comment by gdude2002** [comment]
image

**Review by lemonsaurus — APPROVED** [review]
I think we're good to go.

**Review by jb3 — APPROVED** [review]

File: Pipfile
Comment (lemonsaurus) [reply Pipfile]: This has now been resolved.

**Pull request merged**
